### ISSUE TITLE
av-store does not keep assumed invariant of checked data

### ISSUE THREAD

**Issue by eskimor** [paritytech/polkadot#6076]
We assume in a couple of places that anything stored in the av-store has been checked for validity already, yet the av-store does not guarantee this invariant. Callers ensure that by themselves, e.g. [here](https://github.com/paritytech/polkadot/blob/ffed273daba5e5c66e917c09469e6a042b994ed8/node/network/availability-distribution/src/requester/fetch_task/mod.rs#L303-L309).

Instead the av-store API should either mandate via `newtype` wrappers for example that anything put there has been checked already or check it itself before actually storing/failing if it is found invalid.

Note: There is the related issue of db corruptions: In this case we are able to retrieve invalid data from the av-store, although everything has been checked. Given that this is an unexpected situation, I think this is acceptable (checking also on `load` would resolve that, but I don't think we can justify the performance drawback), but should be clearly documented in the API and we can provide additional /checked/ message variants:

```
Everything retrieved from the av-store is guaranteed to have been checked for validity. In case of hardware faults/db corruptions it is still possible though to retrieve invalid data. Care must therefore be taken if the assumption of validity would have significant consequences if violated due to corruption, e.g. raising a dispute. In such cases, callers are advised to check validity again on retrieval. This can be done by using the `...Checked` variants of query messages.

 Example: availability-recovery: Here we retrieve existing chunks from disk, then fetch missing chunks from validators and try to recover the candidate afterwards - this will fail if the candidate fetched from disk was invalid due to corruption.
```

Additional /Checked/ message variants:

```
AvailabilityStoreMessage::QueryAvailableDataChecked
AvailabilityStoreMessage::QueryChunkChecked
AvailabilityStoreMessage::QueryAllChunksChecked
```

**Comment by eskimor** [comment]
Tasks:

- [ ] Add checks to `Store` messages where applicable
- [ ] Update API documentation to state invariants kept
- [ ] Add `Checked` message types
- [ ] Remove checks at calling sites

**Comment by rphmeier** [comment]
Could we also address this by adding a checksum to DB entries? Comparing the loaded data to the checksum should indicate whether it has been corrupted and would be cheaper than full checks.

**Comment by eskimor** [comment]
Yeah that's nice, so instead of the `checked` variants we would just do a quick checksum check always.
